flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaTransactionLog.java

```diff
@@ -0,0 +1,307 @@
+                        main.getTransactionalIdPrefix());
+                return;
+            }
+
+            final int openSubtaskIndex = openTransaction.getSubtaskId();
+            final long openCheckpointOffset = openTransaction.getTransactionalIdOffset();
+
+            final boolean commonRemainder =
+                    openSubtaskIndex % numberOfParallelSubtasks == main.getSubtaskId()
+                            && openSubtaskIndex != main.getSubtaskId();
+            final boolean sameSubtaskHigherCheckpoint =
+                    hasSameSubtaskWithHigherCheckpoint(openSubtaskIndex, openCheckpointOffset);
+            final boolean higherSubtasksWithFirstCheckpoint =
+                    minOthers.isPresent()
+                            && minOthers.getAsInt() < openSubtaskIndex
+                            && openCheckpointOffset == 0;
+
+            if (!commonRemainder
+                    && !sameSubtaskHigherCheckpoint
+                    && !higherSubtasksWithFirstCheckpoint) {
+                return;
+            }
+            openTransactions.add(transactionalId);
+        };
+    }
+
+    @Override
+    public void close() {
+        consumer.close();
+    }
+
+    private boolean hasSameSubtaskWithHigherCheckpoint(
+            int openSubtaskIndex, long openCheckpointOffset) {
+        return combined.stream()
+                .anyMatch(
+                        state ->
+                                state.getSubtaskId() == openSubtaskIndex
+                                        && state.getTransactionalIdOffset()
+                                                <= openCheckpointOffset);
+    }
+
+    private static boolean isTransactionInFinalState(TransactionState state) {
+        return TERMINAL_TRANSACTION_STATES.contains(state);
+    }
+
+    private static Optional<KafkaWriterState> fromTransactionalId(String transactionalId) {
+        final String[] splits = transactionalId.split(TRANSACTIONAL_ID_DELIMITER);
+        if (splits.length < 3) {
+            LOG.debug("Transaction {} was not created by the Flink Kafka sink", transactionalId);
+            return Optional.empty();
+        }
+        try {
+            final long checkpointOffset = Long.parseLong(splits[splits.length - 1]);
+            final int subtaskId = Integer.parseInt(splits[splits.length - 2]);
+            return Optional.of(
+                    new KafkaWriterState(
+                            String.join(
+                                    TRANSACTIONAL_ID_DELIMITER,
+                                    Arrays.copyOfRange(splits, 0, splits.length - 2)),
+                            subtaskId,
+                            checkpointOffset));
+        } catch (NumberFormatException e) {
+            LOG.debug(
+                    "Transaction {} was not created by the Flink Kafka sink: {}",
+                    transactionalId,
+                    e);
+            return Optional.empty();
+        }
+    }
```

Comment:
Please add unit test, especially to cover a prefix with the delimiter (code looks good) ;) 
Also please avoid the costly `split` and use Guava's Splitter instead.